gk/main.c

```diff
@@ -1739,56 +1811,95 @@ process_pkts_front(uint16_t port_front, uint16_t port_back,
 			continue;
 		}
 
-		ip_flow_hash_val = likely(front->rss) ? pkt->hash.rss :
-			rss_ip_flow_hf(&packet.flow, 0, 0);
+		flow_arr[num_ip_flows] = &packet->flow;
+		flow_hash_val_arr[num_ip_flows] = likely(front->rss) ?
+			pkt->hash.rss : rss_ip_flow_hf(&packet->flow, 0, 0);
+		num_ip_flows++;
+	}
 
-		/* 
-		 * Find the flow entry for the IP pair.
-		 *
-		 * If the pair of source and destination addresses 
-		 * is in the flow table, proceed as the entry instructs,
-		 * and go to the next packet.
-		 */
-		ret = rte_hash_lookup_with_hash(instance->ip_flow_hash_table,
-			&packet.flow, ip_flow_hash_val);
-		if (ret >= 0)
-			fe = &instance->ip_flow_entry_table[ret];
+	done_lookups = 0;
+	while (done_lookups < num_ip_flows) {
+		uint32_t num_keys = num_ip_flows - done_lookups;
+		if (num_keys > RTE_HASH_LOOKUP_BULK_MAX)
+			num_keys = RTE_HASH_LOOKUP_BULK_MAX;
+
+		ret = rte_hash_lookup_bulk_with_hash(
+			instance->ip_flow_hash_table,
+			(const void **)&flow_arr[done_lookups],
+			(hash_sig_t *)&flow_hash_val_arr[done_lookups],
+			num_keys, &pos_arr[done_lookups]);
+		if (ret != 0) {
+			GK_LOG(NOTICE,
+				"failed to find multiple keys in the hash table at lcore %u\n",
+				rte_lcore_id());
+		}
+
+		done_lookups += num_keys;
+	}
+
+	for (i = 0; i < num_ip_flows; i++) {
+		if (pos_arr[i] >= 0)
+			fe_arr[i] = &instance->ip_flow_entry_table[pos_arr[i]];
 		else {
-			/*
-			 * Otherwise, look up the destination address
-		 	 * in the global LPM table.
-			 */
-			struct gk_fib *fib = look_up_fib(
-				&gk_conf->lpm_tbl, &packet.flow);
-
-			fe = lookup_fe_from_lpm(&packet, ip_flow_hash_val,
-				fib, &num_tx, tx_bufs, acl4, acl6, num_pkts,
-				icmp_bufs, req_bufs, req_prio, &num_reqs,
-				front, back, instance, gk_conf);
-			if (fe == NULL)
-				continue;
+			fe_arr[i] = NULL;
+			if (flow_arr[i]->proto == RTE_ETHER_TYPE_IPV4) {
+				lpm_lookup_pos[num_lpm_lookups] = i;
+				flows[num_lpm_lookups] = flow_arr[i];
+				num_lpm_lookups++;
+			} else {
+				lpm6_lookup_pos[num_lpm6_lookups] = i;
+				flows6[num_lpm6_lookups] = flow_arr[i];
+				num_lpm6_lookups++;
+			}
 		}
+	}
+
+	/* The remaining flows need LPM lookups. */
+	lookup_fib_bulk(&gk_conf->lpm_tbl, flows, num_lpm_lookups, fibs);
+	lookup_fib6_bulk(&gk_conf->lpm_tbl, flows6, num_lpm6_lookups, fibs6);
+
+	for (i = 0; i < num_lpm_lookups; i++) {
+		int fidx = lpm_lookup_pos[i];
+
+		fe_arr[fidx] = lookup_fe_from_lpm(&pkt_arr[fidx],
+			flow_hash_val_arr[fidx], fibs[i], &num_tx, tx_bufs,
+			acl4, acl6, num_pkts, icmp_bufs, req_bufs, req_prio,
+			&num_reqs, front, back, instance, gk_conf);
+	}
+
+	for (i = 0; i < num_lpm6_lookups; i++) {
+		int fidx = lpm6_lookup_pos[i];
+
+		fe_arr[fidx] = lookup_fe_from_lpm(&pkt_arr[fidx],
+			flow_hash_val_arr[fidx], fibs6[i], &num_tx, tx_bufs,
+			acl4, acl6, num_pkts, icmp_bufs, req_bufs, req_prio,
+			&num_reqs, front, back, instance, gk_conf);
+	}
+
+	for (i = 0; i < num_ip_flows; i++) {
+		if (fe_arr[i] == NULL)
+			continue;
```

Comment:
`lookup_fe_from_lpm()` must issue an warning when there is no fun to deal with this flow. Moreover, beside the warning, we must free the packet to avoid memory leak and account as a dropped packet in the statistics.

Reply:
When ``fe_arr[i]`` is ``NULL``, the packet will be either dropped or dealt by other blocks within ``lookup_fe_from_lpm()``. Thus, we don't need to free the packet here.

Reply:
You are right. AND we don't need the log message here either since `lookup_fe_from_lpm()` will issue it including the flow information.